packages/flutter/test/widgets/network_image_resolution_test.dart

```diff
@@ -0,0 +1,137 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/services.dart';
+import 'package:flutter/http.dart' as http;
+import 'package:flutter/material.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+class NetworkImageLoadingListenerMock extends NetworkImageLoadingListener {
+
+  bool imageloadingsuccess;
+  bool imageloadingfailed;
+
+  @override
+  void networkImageLoadingSuccess() {
+    this.imageloadingsuccess = true;
+  }
+
+  @override
+  void networkImageLoadingFailed() {
+    this.imageloadingfailed = true;
+  }
+
+  Future<bool> waitLoadingSuccessOrFailed() async {
+    while(this.imageloadingsuccess == null && this.imageloadingfailed == null) { }
+    return true;
+  }
+
+}
+
+void main() {
+
+  testWidgets('network image listener, loading failed', (WidgetTester tester) async {
+
+    NetworkImageLoadingListenerMock listener = new NetworkImageLoadingListenerMock();
+
+    http.Client.clientOverride = () {
+      return new http.MockClient((http.BaseRequest request) {
+        return new Future<http.Response>.value(
+          new http.Response("return an error.", 404, request: request)
+        );
+      });
+    };
+
+    await tester.pumpWidget(
+      new MaterialApp(
+        home: new Material(
+          child: new Builder(
+            builder: (BuildContext context) {
+              return new Center(
+                child: new Image.network('fake_url', networkImageLoadingListener: listener)
+              );
+            }
+          )
+        )
+      )
+    );
+
+    await (listener.waitLoadingSuccessOrFailed());
+    expect(listener.imageloadingsuccess, null);
+    expect(listener.imageloadingfailed, true);
+
+  });
+
+  testWidgets('network image listener, loading success but empty answer', (WidgetTester tester) async {
+
+    NetworkImageLoadingListenerMock listener = new NetworkImageLoadingListenerMock();
+
+    http.Client.clientOverride = () {
+      return new http.MockClient((http.BaseRequest request) {
+        print('bad answer');
+        return new Future<http.Response>.value(
+          new http.Response("", 200, request: request)
```

Comment:
Mmm, interesting answer. Indeed, the ImageNetwork class return a null ImageInfo if the image failed to load from the network, if the network response is empty (byte array length = 0) or if the native method to decode the image failed. 
Should we call the success callback if the network response is OK (200), whatever if the image is correctly decoded or not, or should we call the success callback only if the ImageInfo code != null ? It is a good question. What is your mind?

Another question concerning the cache, if a ImageNetwork fail to load, should we try to reload it the next time a ImageNetwork is called with the same url? 
For now, if the load fail, an ImageNetwork with the same url will never be loaded.
I think we should try to reload it.

Doing this test on the cache, I'v seen that my actual implementation with callbacks is not good :-(
If the ImageNetwork stream has been cached, then the callbacks will never be called.
The solution is not trivial, I can't figure out where to code it. I would really appreciate help :-(